src/main/java/frc/robot/swerve/ModuleIOReal.java

```diff
@@ -0,0 +1,181 @@
+        angleMotorConfig.Feedback.RotorToSensorRatio = 1;
+        angleMotorConfig.MotorOutput.DutyCycleNeutralDeadband = SwerveConstants.NEUTRAL_DEADBAND;
+        angleMotor.getConfigurator().apply(angleMotorConfig);
+
+        angleMotor.setNeutralMode(NeutralModeValue.Brake);
+        angleMotor.setInverted(true);
+    }
+
+    @Override
+    public void updateInputs(SwerveModuleInputs inputs) {
+        inputs.absolutePosition = encoder.getAbsolutePosition();
+
+        inputs.driveMotorSupplyCurrent = driveMotor.getSupplyCurrent().getValue();
+        inputs.driveMotorStatorCurrent = driveMotor.getStatorCurrent().getValue();
+        driveSupplyChargeUsedCoulomb.update(inputs.driveMotorSupplyCurrent);
+        inputs.driveMotorSupplyCurrentOverTime = driveSupplyChargeUsedCoulomb.get();
+        driveStatorChargeUsedCoulomb.update(inputs.driveMotorStatorCurrent);
+        inputs.driveMotorStatorCurrentOverTime = driveStatorChargeUsedCoulomb.get();
+        inputs.driveMotorPosition = driveMotor.getRotorPosition().getValue();
+        inputs.driveMotorVelocity = getVelocity();
+        inputs.driveMotorVelocitySetpoint = driveMotorVelocitySetpoint;
+
+        inputs.angleMotorSupplyCurrent = angleMotor.getSupplyCurrent().getValue();
+        inputs.angleMotorStatorCurrent = angleMotor.getStatorCurrent().getValue();
+        angleSupplyChargeUsedCoulomb.update(inputs.angleMotorSupplyCurrent);
+        inputs.angleMotorSupplyCurrentOverTime = angleSupplyChargeUsedCoulomb.get();
+        angleStatorChargeUsedCoulomb.update(inputs.angleMotorStatorCurrent);
+        inputs.angleMotorStatorCurrentOverTime = angleStatorChargeUsedCoulomb.get();
+        inputs.angleMotorPosition = angleMotor.getRotorPosition().getValue();
+        inputs.angleMotorVelocity = angleMotor.getVelocity().getValue();
+
+        inputs.angle = getAngle();
+        currentAngle = inputs.angle;
+
+        inputs.angleSetpoint = angleSetpoint;
+
+        inputs.moduleDistance = getModulePosition().distanceMeters;
+    }
+
+    @Override
+    public double getAngle() {
+        return AngleUtil.normalize(Units.rotationsToRadians(angleMotor.getPosition().getValue()));
+    }
+
+    @Override
+    public void setAngle(double angle) {
+        angleSetpoint = AngleUtil.normalize(angle);
+        Rotation2d error = new Rotation2d(angle).minus(new Rotation2d(currentAngle));
+        angleMotor.setControl(
+                new MotionMagicVoltage(angleMotor.getPosition().getValue() + error.getRotations())
+                        .withEnableFOC(true));
+    }
+
+    @Override
+    public double getVelocity() {
+        return driveMotor.getVelocity().getValue();
+    }
+
+    @Override
+    public void setVelocity(double velocity) {
+        var angleError = new Rotation2d(angleSetpoint).minus(new Rotation2d(currentAngle));
+        velocity *= angleError.getCos();
+
+        driveMotorVelocitySetpoint = velocity;
```

Comment:
This is the same for simulation and real-world, it should be in the interface/subsystem.

Reply:
But it uses local variables in the classes. putting it in the subsystem would add unnecessary variables.

Reply:
It uses local variables and adding them to the subsystem would make it more crowded for no reason.

Reply:
Why more crowded? I only offer to move code from this classe to another.

Reply:
This is pretty important... for replay to work as expected, the same logic needs to be used for all inmplementations.

Reply:
Oh sorry I didn't see this. I agree I'll fix it.

Reply:
Oh sorry I missed this.
To move this logic to the module class I'll need to create the angle setpoint and the current angle for that class.

Reply:
No worries :)
If I am not mistaken, you can access those from the `inputs` member, like in [6328's 2023 code](https://github.com/Mechanical-Advantage/RobotCode2023/blob/9884d13b2220b76d430e82248fd837adbc4a10bc/src/main/java/org/littletonrobotics/frc2023/subsystems/drive/Module.java#L121-L124).